src/RadixDlt.NetworkGateway.PostgresIntegration/Services/TransactionQuerier.cs

```diff
@@ -322,18 +325,71 @@ public async Task<ICollection<StatusLookupResult>> LookupPendingTransactionsByIn
         return pendingTransactions.Select(pt => new StatusLookupResult(pt.PayloadHash.ToHex(), pt.Status.ToGatewayModel(), pt.LastFailureReason)).ToArray();
     }
 
-    private async Task<List<GatewayModel.CommittedTransactionInfo>> GetTransactions(List<long> transactionStateVersions, GatewayModel.TransactionCommittedDetailsOptIns optIns, CancellationToken token)
+    private async Task<List<GatewayModel.CommittedTransactionInfo>> GetTransactions(
+        List<long> transactionStateVersions, GatewayModel.TransactionCommittedDetailsOptIns optIns, CancellationToken token)
     {
         var transactions = await _dbContext.LedgerTransactions
             .Where(ult => transactionStateVersions.Contains(ult.StateVersion))
             .ToListAsync(token);
 
         var entityIdToAddressMap = await GetEntityAddresses(transactions.SelectMany(x => x.AffectedGlobalEntities).ToList(), token);
 
-        return transactions
-            .OrderBy(lt => transactionStateVersions.IndexOf(lt.StateVersion))
-            .Select(lt => lt.ToGatewayModel(optIns, entityIdToAddressMap))
+        var schemaHashes = transactions
+            .Where(x => x.EngineReceipt.EventsSchemaHash != null)
+            .SelectMany(x => x.EngineReceipt.EventsSchemaHash!)
             .ToList();
+
+        List<PackageSchemaHistory>? schemas = null;
+
+        if (optIns.ReceiptEvents && schemaHashes.Any())
+        {
+            schemas = await _dbContext.PackageSchemaHistory
+                .Where(x => schemaHashes.Contains(x.SchemaHash))
+                .ToListAsync(token);
+        }
+
+        List<GatewayModel.CommittedTransactionInfo> mappedTransactions = new List<GatewayModel.CommittedTransactionInfo>();
+        var networkId = _networkConfigurationProvider.GetNetworkId();
+
+        foreach (var transaction in transactions.OrderBy(lt => transactionStateVersions.IndexOf(lt.StateVersion)))
+        {
+            if (!optIns.ReceiptEvents || schemaHashes?.Any() == false)
+            {
+                mappedTransactions.Add(transaction.ToGatewayModel(optIns, entityIdToAddressMap, null));
+            }
+            else
+            {
+                List<string> events = new List<string>();
+
+                for (var i = 0; i < transaction.EngineReceipt.EventsSbor!.Length; ++i)
```

Comment:
1. Yeah, I'm quite sure we should use non-nullable collections.
2. `++i` seriously? :D
3. I guess we should create some extension method returning a collection of `{sober, schema_hash, type_index, index_kind`.